File: v1/orrbeam/sunshine_api.py
"""Sunshine REST API client for pairing automation."""

import asyncio
import logging
import ssl

import aiohttp

log = logging.getLogger("orrbeam.sunshine_api")


def _ssl_context() -> ssl.SSLContext:
    """SSL context that accepts self-signed certs (standard for Sunshine)."""
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE
    return ctx


async def submit_pin(host: str, port: int, username: str, password: str,
                     pin: str, client_name: str = "orrbeam") -> bool:
    """Submit a pairing PIN to Sunshine's web API.

    Retries because the PIN endpoint only succeeds when a pairing request
    is actively pending (timing window after moonlight pair starts).
    """
    url = f"https://{host}:{port}/api/pin"
    auth = aiohttp.BasicAuth(username, password)

    for attempt in range(15):
        try:
            async with aiohttp.ClientSession() as session:
                async with session.post(
                    url,
                    json={"pin": pin, "name": client_name},
                    auth=auth,
                    ssl=_ssl_context(),
                    timeout=aiohttp.ClientTimeout(total=5),
                ) as resp:
                    if resp.status == 200:
                        data = await resp.json()
                        if data.get("status"):
                            log.info("PIN accepted by Sunshine at %s", host)
                            return True
                        log.debug("PIN attempt %d: status=false (no pending request yet)", attempt + 1)
                    else:
                        log.debug("PIN attempt %d: HTTP %d", attempt + 1, resp.status)
        except Exception as e:
            log.debug("PIN attempt %d: %s", attempt + 1, e)
        await asyncio.sleep(1.0)

    log.warning("PIN submission failed after 15 attempts for %s", host)
    return False

File: crates/orrbeam-core/src/sunshine_api.rs
//! HTTP client for the local Sunshine pairing API.
//!
//! Provides [`submit_pin_local`] for submitting a Moonlight pairing PIN to the
//! Sunshine web API (`https://localhost:47990`).

use serde::{Deserialize, Serialize};
use thiserror::Error;

/// Errors that can occur when communicating with the local Sunshine API.
#[derive(Error, Debug)]
pub enum SunshineApiError {
    /// An HTTP-level error (connection refused, non-2xx response, etc.).
    #[error("HTTP request failed: {0}")]
    Http(String),
    /// Sunshine rejected the PIN after the maximum number of attempts.
    #[error("PIN rejected by Sunshine")]
    PinRejected,
    /// The Sunshine API could not be reached at `https://localhost:47990`.
    #[error("Sunshine API not reachable")]
    Unreachable,
    /// The `sunshine_username` or `sunshine_password` fields are empty.
    #[error("missing Sunshine credentials")]
    NoCredentials,
}

#[derive(Debug, Serialize, Deserialize)]
struct PinRequest {
    pin: String,
    name: String,
}

#[derive(Debug, Deserialize)]
struct PinResponse {
    status: Option<bool>,
}

/// Submit a pairing PIN to the local Sunshine instance.
///
/// Sunshine's web API runs on https://localhost:47990 with self-signed TLS.
/// The PIN endpoint only succeeds when a Moonlight pairing request is actively
/// pending, so we retry up to `max_attempts` times with 1-second intervals.
pub async fn submit_pin_local(
    username: &str,
    password: &str,
    pin: &str,
    client_name: &str,
    max_attempts: u32,
) -> Result<(), SunshineApiError> {
    if username.is_empty() || password.is_empty() {
        return Err(SunshineApiError::NoCredentials);
    }

    let client = reqwest::Client::builder()
        .danger_accept_invalid_certs(true)
        .build()
        .map_err(|e| SunshineApiError::Http(e.to_string()))?;

    let url = "https://127.0.0.1:47990/api/pin";

    for attempt in 1..=max_attempts {
        match client
            .post(url)
            .basic_auth(username, Some(password))
            .json(&PinRequest {
                pin: pin.to_string(),
                name: client_name.to_string(),
            })
            .timeout(std::time::Duration::from_secs(5))
            .send()
            .await
        {
            Ok(resp) if resp.status().is_success() => {
                if let Ok(data) = resp.json::<PinResponse>().await
                    && data.status.unwrap_or(false)
                {
                    tracing::info!("PIN accepted by local Sunshine (attempt {attempt})");
                    return Ok(());
                }
                tracing::debug!("PIN attempt {attempt}/{max_attempts}: not yet accepted");
            }
            Ok(resp) => {
                tracing::debug!(
                    "PIN attempt {attempt}/{max_attempts}: HTTP {}",
                    resp.status()
                );
            }
            Err(e) => {
                tracing::debug!("PIN attempt {attempt}/{max_attempts}: {e}");
            }
        }

        if attempt < max_attempts {
            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        }
    }

    Err(SunshineApiError::PinRejected)
}
